versioned_docs/version-4.0/devupdate.md

```diff
@@ -0,0 +1,958 @@
+
+All course format output classes implements the new `named_templatable` interface, which allows the class to define its own template path using the `get_template_name` method. This new interface in combination with [mustache blocks](https://docs.moodle.org/dev/Templates#Blocks) allows the format plugins to provide alternative templates to render the course.
+
+:::tip Migrating 3.11 course formats doc
+
+All the new output classes and a guide on how to migrate the current third-party plugins can be found in the [Migrating 3.11 formats](/docs/apis/plugintypes/format/migration) page.
+
+:::
+
+### Course editor javascript modules and frontend components
+
+The majority of the javascript logic related to the course editing is replaced by AMD modules. Because this is a major change in the way courses are edited and rendered, by default format plugins will continue using the previous YUI modules for now. However, formats can start using the new libraries overriding the `$format->supports_components()` method.
+
+Some Moodle 4.0 new features are only available for courses using the new editor library:
+
+- Edit the course via the course index
+- Creating sections without reloading the course page
+- The new move section/activity modal
+- Native browser drag&drop implementation
+
+The new course editor uses a component-based reactive pattern to keep track of the course changes. The pattern highlights are:
+
+- The main AMD module `core_crouseformat\courseeditor` maintains a data structure called state.
+- Each UI element is implemented as a Component that observes the course state data and reacts to any data change
+- When any reactive component needs to modify the course, it asks the course editor to execute a mutation. Mutations encapsulate all web services calls and alter the course state data.
+
+:::tip Reactive library doc
+
+The reactive library documentation, as well as the format plugin migration guide, can be found in the [Moodle 3.11 vs 4.0 course editor architecture](/docs/guides/javascript/reactive) page.
+
+:::
+
+### Other course related 4.0 changes
+
+Two new web services have been added:
+
+- `core_courseformat_get_state`: user by the new javascript course editor to get the current course state data (containing the list of sections, activities, and other course-related data)
+- `core_courseformat_update_course`: to alter the current course content. Each call returns the parts of the course state altered by the action
+
+## Behat changes
+
+### New steps
+
+Moodle 4.0 introduces some new behat steps.
+
+Sometimes you want to create a bulk number of activities. In that case you can use:
+
+```
+:count :entitytype exist with the following data:
+```
+
+For example:
+
+```gherkin
+Given 100 "mod_lti > tool types" exist with the following data:
+  |name        |Test tool [](count)                  |
+  |description |Example description [](count)        |
+  |baseurl     |https://www.example.com/tool[](count)|
+
+```
```

Comment:
```suggestion
```gherkin
Given 100 "mod_lti > tool types" exist with the following data:
  | name        | Test tool [](count)                   |
  | description | Example description [](count)         |
  | baseurl     | https://www.example.com/tool[](count) |

```
```

Reply:
I will remove the empty line but will keep the closing blockquote ;-)

Reply:
Good plan!